### PR TITLE
[WIP] Support for images

### PR DESCRIPTION
**Pull request by RuiGuilherme**
Features: 

- Change history size: **MAX** 2000 and **MIN** 2
- Support to save images **(Optional)**
- Image preview on menu https://github.com/Tudmotu/gnome-shell-extension-clipboard-indicator/pull/286/commits/d6475bf95cce05130b4460e3d920bcf08e038664
- Support for image/png **(Optional)**
- Support to file manager (Tested in **Dolphin**, **Nautilus** and **Nemo**) https://github.com/Tudmotu/gnome-shell-extension-clipboard-indicator/commit/7a97da53d0c25aedaf503d1cf2a4cc3831a60abc

Know issue:

- Large images (9 mb or more) may not save. (8k wallpaper for example haha)
- image/png only works with Mutter 41.3+ ([patch mutter](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1812) )

### CODE REVIEW

**Comment by RuiGuilherme** [comment]
Probably related:
#112 
#70 
#222  
#154 
#21

**Comment by Tudmotu** [comment]
Hi :) 
Thank you for this, I hope you can make this work as this is a feature I personally have been waiting for.

I see you are serializing images as Base64 and save it inside the json, right?
What do you think about saving the images on the file system, and just reference their path in the JSON?

**Comment by RuiGuilherme** [comment]
> I see you are serializing images as Base64 and save it inside the json, right?
Yes, but for _addEntry() I use bytes because for some reason convert Base64 to Bytes inside Clipboard.set_content() costs me ~2 Seconds for small imagens.

> What do you think about saving the images on the file system, and just reference their path in the JSON?
I will try using JSON, that should probably avoid the "throttling".

**Comment by RuiGuilherme** [comment]
[commit bc45597](https://github.com/Tudmotu/gnome-shell-extension-clipboard-indicator/pull/286/commits/bc455972e9ca5f78711818db90be6b952ea62aef) now save files inside .cache and JSON save only path... 

I did several tests with and without the mutter patch... All works fine.

### Fix:
- 1mb images or more large can cause a "throttling".
- many images in history also cause "throttling".
- can cause delay on gnome start and restart


### Know issue:
- image/png works only with this [patch mutter](https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1812)
- Large images (9 mb or more) may not save. (8k wallpaper for example haha)

**Comment by Tudmotu** [comment]
Hi @RuiGuilherme :) 
Is this still WIP?

**Comment by RuiGuilherme** [comment]
Yup, I need more testers... it works well on my i7 9750h but I don't have low-end spec to test possibles issues performance.

I will try to find someone available in some group related to GNOME to test.

File: utils.js [hunk 1]
```diff
@@ -14,6 +14,40 @@ const REGISTRY_FILE = 'registry.txt';
...
+            const stream = obj.replace_finish(res);
+
+            stream.write_bytes_async(buffer, GLib.PRIORITY_DEFAULT, null, function (w_obj, w_res) {
+                w_obj.write_bytes_finish(w_res);
+                stream.close(null);
+            });
+        });
+    }
+
+    return path;
```
Comment (tcanabrava) [inline utils.js:38]: since you are .create_async() the file here, are you sure that the file will exist when this function returns?

**Comment by canadaduane** [comment]
I gave this a spin, but I think there is a bug when copying regular text:

1. With a fresh install/cache, copy some text
2. View the Clipboard Indicator history via clicking top bar icon (drop-down)
3. Copied text appears as a filepath, rather than the copied text itself

The filepath also appears to be incorrect: `/home/duane/.cache/clipboard-indicator@tudmotu.com/undefined`

File: extension.js [hunk 2]
```diff
@@ -32,9 +33,13 @@ const Me = ExtensionUtils.getCurrentExtension();
 const Utils = Me.imports.utils;
 const ConfirmDialog = Me.imports.confirmDialog;
 const Prefs = Me.imports.prefs;
+
 const prettyPrint = Utils.prettyPrint;
 const writeRegistry = Utils.writeRegistry;
 const readRegistry = Utils.readRegistry;
+const writerImage = Utils.writerImage;
```
Comment (canadaduane) [inline extension.js:40]: Should this be called `writeImage`?

File: extension.js [hunk 3]
```diff
@@ -466,26 +514,71 @@ const ClipboardIndicator = Lang.Class({
...
 
         Clipboard.get_text(CLIPBOARD_TYPE, function (clipBoard, text) {
-            that._processClipboardContent(text);
+            if (text === null && SAVE_IMAGE) {
+                Clipboard.get_content(CLIPBOARD_TYPE, SUPPORT_PNG ? 'image/png' : 'image/jpeg', function (clipboard, content) {
+                    that._processClipboardContent(content.get_data(), 'image', content.hash());
+                })
+            }
+            else {
+                const itens = text.split('\n');
```
Comment (canadaduane) [inline extension.js:523]: I think `text` can be null here. Consider:

1. let `text` be null
2. let `SAVE_IMAGE` be false

In that case, we fall through to the `else` block on line 522, and this line will fail when it tries to split a null value.

Also, as an aside: should this variable (`const itens`) be named `items`?

File: extension.js [hunk 4]
```diff
@@ -466,26 +514,71 @@ const ClipboardIndicator = Lang.Class({
...
+        }
+        else {
+            content = writerImage(byteContent, name);
         }
 
-        if (text !== "" && text) {
+        if (content !== "" && content) {
             let registry = that.clipItemsRadioGroup.map(function (menuItem) {
-                return menuItem.clipContents;
+                return menuItem.clipContents
```
Comment (canadaduane) [inline extension.js:576]: Leave semi-colon here?

File: extension.js [hunk 5]
```diff
@@ -494,24 +587,24 @@ const ClipboardIndicator = Lang.Class({
...
-        this._addEntry(item.clipContents, item.clipFavorite, item.currentlySelected, false);
+        this._addEntry(item.clipContents, item.clipFavorite, item.currentlySelected, false, type ?? item.type);
     },
 
-    _findItem: function (text) {
-        return this.clipItemsRadioGroup.filter(
-            item => item.clipContents === text)[0];
+    _findItem: function (content, type) {
+            return this.clipItemsRadioGroup.filter(
+                item => item.clipContents === content)[0];
```
Comment (canadaduane) [inline extension.js:607]: Formatting--looks like extra tabs were added here?

File: extension.js [hunk 6]
```diff
@@ -682,6 +775,8 @@ const ClipboardIndicator = Lang.Class({
         TOPBAR_DISPLAY_MODE  = this._settings.get_int(Prefs.Fields.TOPBAR_DISPLAY_MODE_ID);
         DISABLE_DOWN_ARROW   = this._settings.get_boolean(Prefs.Fields.DISABLE_DOWN_ARROW);
         STRIP_TEXT           = this._settings.get_boolean(Prefs.Fields.STRIP_TEXT);
+        SAVE_IMAGE           = this._settings.get_boolean(Prefs.Fields.SAVE_IMAGE);
+        SUPPORT_PNG           = this._settings.get_boolean(Prefs.Fields.SUPPORT_PNG);
```
Comment (canadaduane) [inline extension.js:779]: Line up equals signs?

File: extension.js [hunk 7]
```diff
@@ -466,26 +514,71 @@ const ClipboardIndicator = Lang.Class({
...
 
-    _processClipboardContent (text) {
+    // type most be image or text
+    _processClipboardContent (byteContent, type, name) {
         const that = this;
-
-        if (STRIP_TEXT) {
-            text = text.trim();
+        let content;
+        if (STRIP_TEXT && type === "text") {
```
Comment (canadaduane) [inline extension.js:567]: I think we're missing the case where STRIP_TEXT is false here. Shouldn't `content` be set to `byteContent` regardless?

Comment (canadaduane) [reply extension.js:567]: e.g. ```        if (type === "text") {
            if (!STRIP_TEXT) {
                content = byteContent;
            }
            else {
                content = byteContent.trim();
            }
        }
        else {
            content = writerImage(byteContent, name);
        }
```

**Comment by canadaduane** [comment]
I've made a PR to @RuiGuilherme's branch https://github.com/RuiGuilherme/gnome-shell-extension-clipboard-indicator/pull/3

(My first time making a PR to a PR!)

**Comment by canadaduane** [comment]
FWIW, this works for me now. Thanks @RuiGuilherme for all your work!

**Comment by canadaduane** [comment]
@Tudmotu can this be merged in do you think?

**Comment by phamthainb** [comment]
This is so good, why is it not merged yet :clap:

**Comment by Tudmotu** [comment]
Hi all,
Am I missing something - when I right-click save image, it does not get added to the clipboard. Does this happen to you as well?

**Comment by Tudmotu** [comment]
Also, ctrl+v doesn't seem to paste images when I select them from the clipboard.

**Comment by Tudmotu** [comment]
I'm using GNOME 41 btw.

**Comment by RuiGuilherme** [comment]
> Hi all, Am I missing something - when I right-click save image, it does not get added to the clipboard. Does this happen to you as well?

I thought some modification of version 41 caused the bugs, so I made a new branch to test and everything works. 
https://youtu.be/r23P8MXqSx0

https://github.com/RuiGuilherme/gnome-shell-extension-clipboard-indicator/tree/Gnome41

**Comment by Tudmotu** [comment]
Ok, interesting. Thanks :)

**Comment by Tudmotu** [comment]
I made some small change due to conflicts and backwards compatibility, maybe I introduced a bug. Will look into it.

**Comment by hyuri** [comment]
Great features guys! I've been waiting for this!
Just wanted to chime in to ask if saving the images on disk could pose a potential privacy issue. In case, for example, that an outside agent gained access to the folder you're storing the images in.

**Comment by RuiGuilherme** [comment]
> Great features guys! I've been waiting for this! Just wanted to chime in to ask if saving the images on disk could pose a potential privacy issue. In case, for example, that an outside agent gained access to the folder you're storing the images in.

Yup, this can happen. However, this is not a specific problem of this feature, a malicious application could access JSON where the texts are stored and have access to all user history. 

But I believe this is not exactly something to worry about, since a malicious application could have access to the clipboard.

**Comment by ilazaridis** [comment]
I am also looking forward to this one. Image support would be really cool! :)

**Comment by mpetuska** [comment]
Just want to put this on the table - GPaste has supported images for a while now and it works quite well. It might be good to dig through their code for ideas?

File: prefs.js [hunk 8]
```diff
@@ -179,6 +183,17 @@ const App = new Lang.Class({
             hexpand: true,
             halign: Gtk.Align.START
         });
+        let saveImageLabel = new Gtk.Label({
+            label: _("Save images on history"),
+            hexpand: true,
+            halign: Gtk.Align.START
+        });
+        // need while this MR is pending: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/1812
+        let pngImageLabel = new Gtk.Label({
```
Comment (SUPERCILEX) [inline prefs.js:192]: That PR was merged, is the workaround still necessary? (Assuming people update to the latest mutter.)

**Review by SUPERCILEX — COMMENTED** [review]
I think this PR could benefit from a fundamentally different approach: instead of special casing images, why not support all file types? This would let us support random stuff like copied google slides, etc. If we don't understand the file type, we simply show its file path (maybe in *italics* to indicate that the path string is not what the user copied). If the path isn't available, we just show the mime type.

In a bit more detail, the idea looks like this:
- Accept any arbitrary clipboard contents
- Add `Save non-text clipboard items to disk` option (disabled by default)
  - If this option is true, we write the contents into our clipboard registry and cache dir
  - If not, we only keep the clipboard contents in memory so as to not waste disk space/cycles
- If the contents is a supported image, we display it in the clipboard UI

From a PR perspective, this offers some nice splits. First write a PR that supports arbitrary clipboard contents in memory. Then write a PR that displays known types (images). Then write a PR that adds the option to save stuff to disk.

**Comment by SUPERCILEX** [comment]
I just tried this out and several things seem broken:
- My clipboard history was nuked, so backwards compat is not being handled properly
- The settings, private mode, and clear history buttons disappeared
- Copying an image from the web did not work (right click, copy image)
- Copying an image from the file manager *did* work, but the aspect ratio is broken and the image is too small
- After selecting a previously copied image, I am unable to paste it for whatever reason

Screenshot:
image

**Comment by Tudmotu** [comment]
Thank you @SUPERCILEX for confirming the issues I encountered.
I opened a new PR that includes backward-compat fixes:
https://github.com/Tudmotu/gnome-shell-extension-clipboard-indicator/pull/334/files

To keep working on this, need to clone the `feature/cache-images` branch and to propose changes need to open a PR that points to `feature/cache-images` branch and not `master`.

**Comment by RuiGuilherme** [comment]
> I just tried this out and several things seem broken:
> 
>     * My clipboard history was nuked, so backwards compat is not being handled properly
> 
>     * The settings, private mode, and clear history buttons disappeared
> 
>     * Copying an image from the web did not work (right click, copy image)
> 
>     * Copying an image from the file manager   _did  _ work, but the aspect ratio is broken and the image is too small
> 
>     * After selecting a previously copied image, I am unable to paste it for whatever reason
> 
> 
> Screenshot: image

Thank you. I'll work on the fixes. 
Did you use Branch for version 40+? 
https://github.com/RuiGuilherme/gnome-shell-extension-clipboard-indicator/tree/Gnome41

Captura de tela de 2022-01-14 11-42-51

**Comment by rasyidrafi** [comment]
Why did this doesnt merged to master branch

**Comment by Tudmotu** [comment]
@rasyidrafi unfortunately this doesn't work.

**Comment by Tudmotu** [comment]
Closing. In the upcoming GNOME 45 version there will be image support (finally lmao).
https://github.com/Tudmotu/gnome-shell-extension-clipboard-indicator/tree/image-support

**Pull request closed without merging**